opencog/atoms/core/RewriteLink.cc

```diff
@@ -88,82 +80,109 @@ inline HandleSeq append_rand_str(const HandleSeq& vars)
 	return new_vars;
 }
 
-Handle RewriteLink::alpha_conversion() const
+Handle RewriteLink::alpha_convert() const
 {
 	HandleSeq vars = append_rand_str(_varlist.varseq);
-	return alpha_conversion(vars);
+	return alpha_convert(vars);
 }
 
-Handle RewriteLink::alpha_conversion(const HandleSeq& vars) const
+Handle RewriteLink::alpha_convert(const HandleSeq& vars) const
 {
 	// Perform alpha conversion
 	HandleSeq hs;
 	for (size_t i = 0; i < get_arity(); ++i)
-		hs.push_back(_varlist.substitute_nocheck(getOutgoingAtom(i), vars));
+		hs.push_back(_varlist.substitute_nocheck(getOutgoingAtom(i), vars, _silent));
 
 	// Create the alpha converted scope link
 	return createLink(hs, get_type());
 }
 
-Handle RewriteLink::alpha_conversion(const HandleMap& vsmap) const
+Handle RewriteLink::alpha_convert(const HandleMap& vsmap) const
 {
 	HandleSeq vars;
 	for (const Handle& var : _varlist.varseq) {
 		auto it = vsmap.find(var);
 		vars.push_back(it == vsmap.end() ? append_rand_str(var) : it->second);
 	}
-	return alpha_conversion(vars);
+	return alpha_convert(vars);
 }
 
 /* ================================================================= */
 
-Handle RewriteLink::partial_substitute(const HandleMap& vm) const
+Handle RewriteLink::beta_reduce(const HandleMap& vm) const
 {
 	// Perform substitution over the variable declaration
-	Handle nvardecl = partial_substitute_vardecl(vm);
+	Handle nvardecl = substitute_vardecl(vm);
 
 	// Perform substitution over the bodies
-	HandleSeq hs = partial_substitute_bodies(nvardecl, vm);
+	HandleSeq hs = substitute_bodies(nvardecl, vm);
 
 	// Filter vardecl
 	nvardecl = filter_vardecl(nvardecl, hs);
 
-	// Insert vardecl in the outgoings if defined
+	// Insert vardecl in the outgoing set, if defined
 	if (nvardecl)
+	{
 		hs.insert(hs.begin(), nvardecl);
+	}
+	else
+	{
+		// Its illegal to create a PutLink that is not in the
+		// form of a redex, i.e. doesn't have variable declarations
+		// in it. So we must not call createLink(), below.
```

Comment:
I don't think it's correct, that is because no vardecl doesn't mean no variable. As you can see https://github.com/opencog/atomspace/blob/master/opencog/atoms/core/RewriteLink.cc#L187 if the RewriteLink didn't have any vardecl to begin with then the resulting vardecl will be `Handle::UNDEFINED`.

I realize that this conflates with https://github.com/opencog/atomspace/blob/master/opencog/atoms/core/RewriteLink.cc#L224 i.e. it shouldn't return `Handle::UNDEFINED` when no variable remains after substitution, it should instead return something like an empty `VariableList`. I'll try to create unit tests for that...

Reply:
Well, but I can guarantee that PutLink will throw. PutLink is a "beta-redex" (google it) and after reducing a beta-redex, it can't continue to be a beta-redex any more, its "gone". 

A lambda is a little more arguable. In standard lambda-calc, after reduction, the lambda is gone.  You can define a non-standard one, which looks like a "function taking zero arguments"  but then the only thing one can do with that is to delta-reduce it.

A pattern with zero variables is ... a toss-up.I guess its well-defined.

Reply:
Got it.